File: ts/lib/src/typescript/Redactor.ts
import * as feistel from 'feistel-cipher'

import { Dictionary } from './Dictionary'

export const DEFAULT_TAG = '~'

export interface Redactor {
  dictionary: Dictionary
  tag: string
  cipher: feistel.FPECipher
  both: boolean

  redact: (line: string, ...delimiters: ReadonlyArray<string>) => string
  expand: (line: string) => string
  clean: (str: string) => string
}

export const redact = (dictionary: Dictionary, tag: string, cipher: feistel.FPECipher, both: boolean) => (line: string, ...delimiters: ReadonlyArray<string>): string =>
  line.split(new RegExp('\\s' + delimiters.map(delim => delim !== ' ' && delim !== '\t' && delim !== '\n' && delim !== '\\s' ? delim : '').join(''))).map(word =>
    both || !dictionary.isEmpty()
      ? dictionary.contains(word)
        ? cipher.encrypt(word)
        : tag !== '' && word.startsWith(tag)
          ? tag + cipher.encrypt(word.substr(tag.length))
          : word
      : !both && dictionary.isEmpty() && tag !== '' && word.startsWith(tag)
        ? tag + cipher.encrypt(word.substr(tag.length))
        : word
  ).join(' ')

export const expand = (dictionary: Dictionary, tag: string, cipher: feistel.FPECipher, both: boolean) => (line: string): string =>
  line.split(' ').map(word =>
    both || !dictionary.isEmpty()
      ? tag !== '' && word.startsWith(tag)
        ? tag + cipher.decrypt(word.substr(tag.length))
        : dictionary.contains(cipher.decrypt(word))
          ? cipher.decrypt(word)
          : word
      : !both && dictionary.isEmpty()
        ? tag !== '' && word.startsWith(tag)
          ? tag + cipher.decrypt(word.substr(tag.length))
          : word
        : word
  ).join(' ')

export const clean = (tag: string) => (str: string): string =>
  tag !== ''
    ? str.replace(new RegExp(tag, 'g'), '')
    : str

export const Redactor = (dictionary: Dictionary, tag: string, cipher: feistel.FPECipher, both: boolean): Redactor => ({
  dictionary,
  tag,
  cipher,
  both: both || !dictionary.isEmpty() && tag !== '',
  redact: redact(dictionary, tag, cipher, both),
  expand: expand(dictionary, tag, cipher, both),
  clean: clean(tag)
})

export const RedactorWithDictionary = (dictionary: Dictionary, cipher: feistel.FPECipher): Redactor =>
  Redactor(dictionary, '', cipher, false)

export const RedactorWithTag = (tag: string, cipher: feistel.FPECipher): Redactor =>
  Redactor(Dictionary([]), tag, cipher, false)

export const DefaultRedactor = (cipher: feistel.FPECipher): Redactor =>
  Redactor(Dictionary([]), DEFAULT_TAG, cipher, false)

File: py/src/redacted/redactor.py
import re
from feistel import FPECipher


from redacted import Dictionary, DEFAULT_TAG


class Redactor:
    """
    Parent class for redactors
    """

    def __init__(self, dictionary: Dictionary, tag: str, cipher: FPECipher, both: bool):
        self.dictionary = dictionary
        self.tag = tag
        self.cipher = cipher
        self.both = both

    def redact(self, line: str, *delimiters) -> str:
        """
        Returns the ciphered version of the passed input data
        """
        actual_delimiters = "\\s" + "".join(
            [
                (
                    delim
                    if delim != " "
                    and delim != "\t"
                    and delim != "\n"
                    and delim != "\\s"
                    else ""
                )
                for delim in delimiters
            ]
        )
        words = re.split(actual_delimiters, line)
        tokens = list[str]()
        for word in words:
            if self.both or not self.dictionary.is_empty():
                if self.dictionary.contains(word):
                    tokens.append(self.cipher.encrypt(word))
                else:
                    if self.tag and word.startswith(self.tag):
                        tokens.append(
                            self.tag + self.cipher.encrypt(word[len(self.tag) :])
                        )
                    else:
                        tokens.append(word)
            else:
                if (
                    not self.both
                    and self.dictionary.is_empty()
                    and self.tag
                    and word.startswith(self.tag)
                ):
                    tokens.append(self.tag + self.cipher.encrypt(word[len(self.tag) :]))
                else:
                    tokens.append(word)

        return " ".join(tokens)

    def expand(self, line: str) -> str:
        """
        Returns the deciphered version of the passed input data
        """
        words = list[str]()
        for word in line.split(" "):
            if self.both or not self.dictionary.is_empty():
                if self.tag and word.startswith(self.tag):
                    words.append(self.tag + self.cipher.decrypt(word[len(self.tag) :]))
                else:
                    if not word.strip():
                        # Avoid using new lines
                        continue
                    else:
                        deciphered = self.cipher.decrypt(word)
                        if deciphered and self.dictionary.contains(deciphered):
                            words.append(deciphered)
                        else:
                            words.append(word)
            else:
                if not self.both and self.dictionary.is_empty():
                    if self.tag and word.startswith(self.tag):
                        words.append(
                            self.tag + self.cipher.decrypt(word[len(self.tag) :])
                        )
                    else:
                        words.append(word)
                else:
                    words.append(word)

        return " ".join(words)

    def clean(self, string: str) -> str:
        """
        Removes any delimiters from the passed string
        """
        if self.tag:
            return string.replace(self.tag, "")
        else:
            return string


class DefaultRedactor(Redactor):
    def __init__(self, cipher: FPECipher):
        super().__init__(Dictionary([]), DEFAULT_TAG, cipher, False)


class RedactorWithDictionary(Redactor):
    def __init__(self, dictionary: Dictionary, cipher: FPECipher):
        super().__init__(dictionary, "", cipher, False)


class RedactorWithTag(Redactor):
    def __init__(self, tag: str, cipher: FPECipher):
        super().__init__(Dictionary([]), tag, cipher, False)
